exonum/src/helpers/fabric/password.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module contains utilities for passphrase entry.
+
+use failure;
+use rpassword::read_password_from_tty;
+
+use std::{env, io, str::FromStr};
+
+use helpers::ZeroizeOnDrop;
+
+/// Passphrase input methods
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum PassInputMethod {
+    /// Prompt passphrase from terminal.
+    Terminal,
+    /// Get passphrase from environment variable (default if `None`).
+    EnvVariable(Option<String>),
+    /// Passphrase is passed as a command line parameter.
+    CmdLineParameter(ZeroizeOnDrop<String>),
+}
+
+/// Secret key types.
+#[derive(Copy, Clone)]
+pub enum SecretKeyType {
+    Consensus,
+    Service,
+}
+
+impl PassInputMethod {
+    /// Get passphrase using selected method.
+    /// Details of this process differs for different secret key types and whether we run node
+    /// or generate config files.
+    pub fn get_passphrase(self, key_type: SecretKeyType, node_run: bool) -> ZeroizeOnDrop<String> {
+        match self {
+            PassInputMethod::Terminal => {
+                let prompt = match key_type {
+                    SecretKeyType::Consensus => "Enter consensus key passphrase",
+                    SecretKeyType::Service => "Enter service key passphrase",
+                };
+                prompt_passphrase(prompt, node_run).expect("Failed to read password from stdin")
+            }
+            PassInputMethod::EnvVariable(name) => {
+                let var = if let Some(ref name) = name {
+                    name
+                } else {
+                    match key_type {
+                        SecretKeyType::Consensus => "EXONUM_CONSENSUS_PASS",
+                        SecretKeyType::Service => "EXONUM_SERVICE_PASS",
+                    }
+                };
+                ZeroizeOnDrop(env::var(var).unwrap_or_else(|e| {
+                    panic!("Failed to get password from env variable: {}, {}", var, e)
+                }))
+            }
+            PassInputMethod::CmdLineParameter(pass) => pass,
+        }
+    }
+}
+
+impl Default for PassInputMethod {
+    fn default() -> Self {
+        PassInputMethod::Terminal
+    }
+}
+
+impl FromStr for PassInputMethod {
+    type Err = failure::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "" {
```

Comment:
Why not `s.is_empty()`?

Reply:
I preferred this variant.
Fine, changed.